### PR TITLE
feat: Edge V2 migration opt-in, capacity budget for migration

### PR DESCRIPTION
**Pull request by khvn26**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [ ] I have added information to `docs/` if required so people know about the feature!
- [X] I have filled in the "Changes" section below?
- [X] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request
- [X] I have filled in the "How did you test this code" section below?

### Naming

`IdentityOverridesV2MigrationStatus` enumeration is renamed to `EdgeV2MigrationStatus`.

### Settings

Added:
- `EDGE_V2_MIGRATION_READ_CAPACITY_BUDGET`: controls how much read capacity is allowed during migration to `environments_v2`. Default value is `0`. The idea is to increase it to a sensible amount whenever Edge V2 is live. 

### Business logic 

Renamed:
- `Project.identity_overrides_v2_migration_status` -> `Project.edge_v2_migration_status`.

Modified:
- `DynamoIdentityWrapper.iter_all_items_paginated` now accepts `capacity_budget` argument. If the budget is reached or exceeded during iteration, it raises a `CapacityBudgetExceeded` exception. If `capacity_budget` is `0`, the error is raised before any request to DynamoDB is made. 

Added: 
- `Project.edge_v2_read_capacity_budget` overrides the `EDGE_V2_MIGRATION_READ_CAPACITY_BUDGET` for the given project. Field value is `null` by default.
- `EdgeV2MigrationStatus.INCOMPLETE` marks projects that have exceeded the read capacity budget during migration and have only environment documents migrated. Identity override migration is completely skipped for such projects.
 
### API 
- Added `edge_v2_migration_status` to `project` endpoints so the migration can be triggered via REST API.
- `edge_v2_read_capacity_budget` is only modifiable via Django admin page. 

## How did you test this code?

All code covered with unit tests.

### CODE REVIEW

File: api/environments/dynamodb/services.py [hunk 1]
```diff
@@ -26,39 +45,55 @@ def migrate_environments_to_v2(project_id: int) -> IdentityOverridesV2Changeset
...
-        "Retrieved %d identity overrides to migrate for project %d",
-        len(changeset.to_put),
-        project_id,
-    )
+    identity_overrides_changeset = IdentityOverridesV2Changeset(to_put=[], to_delete=[])
+    result_status = EdgeV2MigrationStatus.COMPLETE
+
+    try:
+        to_put = list(
+            _iter_paginated_overrides(
```
Comment (gagantrivedi) [inline api/environments/dynamodb/services.py:57]: We should use project_expression to   only fetch the data we need for migration

File: api/environments/dynamodb/services.py [hunk 2]
```diff
@@ -26,39 +45,55 @@ def migrate_environments_to_v2(project_id: int) -> IdentityOverridesV2Changeset
...
-    logger.info(
-        "Retrieved %d identity overrides to migrate for project %d",
-        len(changeset.to_put),
-        project_id,
-    )
+    identity_overrides_changeset = IdentityOverridesV2Changeset(to_put=[], to_delete=[])
+    result_status = EdgeV2MigrationStatus.COMPLETE
+
+    try:
+        to_put = list(
```
Comment (gagantrivedi) [inline api/environments/dynamodb/services.py:56]: Do we have a rough idea about how much data will this be for an environment with several million identities? I just want to make sure we don't hurt task processor

**Review by gagantrivedi — CHANGES_REQUESTED** [review]

File: api/environments/dynamodb/services.py (hunk 1 above)
Comment (khvn26) [reply api/environments/dynamodb/services.py:57]: Added.

File: api/environments/dynamodb/services.py (hunk 2 above)
Comment (khvn26) [reply api/environments/dynamodb/services.py:56]: `len(to_put)` is equal to the count of only the identities that have overridden feature states. The rest of identities is iterated over but discarded.

File: api/environments/dynamodb/wrappers/identity_wrapper.py [hunk 3]
```diff
@@ -89,37 +91,49 @@ def get_all_items(
...
         }
-        if projection_expression:
-            query_kwargs["ProjectionExpression"] = projection_expression
-
         if start_key:
             query_kwargs["ExclusiveStartKey"] = start_key
+        if projection_expression:
+            query_kwargs["ProjectionExpression"] = projection_expression
+        if return_consumed_capacity:
+            query_kwargs["ReturnConsumedCapacity"] = "TOTAL"
```
Comment (matthewelwell) [inline api/environments/dynamodb/wrappers/identity_wrapper.py]: This seems a bit cryptic, I'd love a comment here.

File: api/projects/migrations/0023_rename_identity_overrides_v2_migration_status_project_edge_v2_migration_status.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        ("projects", "0022_add_stale_flags_threshold_to_project"),
+    ]
+
+    operations = [
+        migrations.RenameField(
+            model_name="project",
+            old_name="identity_overrides_v2_migration_status",
+            new_name="edge_v2_migration_status",
+        ),
+    ]
```
Comment (matthewelwell) [inline api/projects/migrations/0023_rename_identity_overrides_v2_migration_status_project_edge_v2_migration_status.py]: Erk, we'll get schema mismatch issues here during deployment! We'll probably have to just rename the python attribute and set the `column_name` on the field definition to keep it the same as it is currently.

**Review by matthewelwell — CHANGES_REQUESTED** [review]
I've gone through the code for now (still need to review the tests) but wanted to submit this now (and go get some dinner...)

File: api/environments/dynamodb/services.py (hunk 2 above)
Comment (gagantrivedi) [reply api/environments/dynamodb/services.py:56]: But are we still fetching all the identities? I thought we found a way to only fetch identities that have overrides?

Comment (khvn26) [reply api/environments/dynamodb/services.py:56]: No, we haven't. Even if we use a filter expression we'll spend the same amount of read capacity.

Comment (gagantrivedi) [reply api/environments/dynamodb/services.py:56]: I think the following should work?

```
 query_kwargs = {
        "IndexName": "environment_api_key-identifier-index",
        "KeyConditionExpression": filter_expression,
        "FilterExpression": Attr("identity_features").ne([]),
    }
``

File: api/tests/unit/environments/dynamodb/test_unit_services.py [hunk 5]
```diff
@@ -73,8 +81,56 @@ def test_migrate_environments_to_v2__wrapper_disabled__does_not_write(
...
+    def iter_all_items_paginated_gen_mock(**_):
+        yield map_identity_to_identity_document(identity)
+        raise CapacityBudgetExceeded(expected_capacity_budget, Decimal(13))
+
+    mocked_dynamodb_identity_wrapper.iter_all_items_paginated.side_effect = (
+        iter_all_items_paginated_gen_mock
+    )
+    mocker.patch(
+        "environments.dynamodb.services.DynamoIdentityWrapper",
+        autospec=True,
```
Comment (matthewelwell) [inline api/tests/unit/environments/dynamodb/test_unit_services.py:113]: minor: it's odd that we use `autospec` here as well as defining the `spec` in the MagicMock constructor above.

File: api/environments/dynamodb/services.py (hunk 2 above)
Comment (khvn26) [reply api/environments/dynamodb/services.py:56]: Let me compare performance between the two variants. If there's no major difference, I'm in favour of this change.

File: api/projects/migrations/0023_rename_identity_overrides_v2_migration_status_project_edge_v2_migration_status.py (hunk 4 above)
Comment (khvn26) [reply api/projects/migrations/0023_rename_identity_overrides_v2_migration_status_project_edge_v2_migration_status.py]: Done.

File: api/environments/dynamodb/wrappers/identity_wrapper.py (hunk 3 above)
Comment (khvn26) [reply api/environments/dynamodb/wrappers/identity_wrapper.py]: Added.

File: api/environments/dynamodb/services.py (hunk 2 above)
Comment (khvn26) [reply api/environments/dynamodb/services.py:56]: Done.

File: api/tests/unit/environments/dynamodb/test_unit_services.py (hunk 5 above)
Comment (khvn26) [reply api/tests/unit/environments/dynamodb/test_unit_services.py:113]: I agree it's a bit overprotective, but it doesn't hurt.

**Review by matthewelwell — APPROVED** [review]

**Review by gagantrivedi — APPROVED** [review]

**Pull request merged**